lambdas/services/fhir_document_reference_service_base.py

```diff
@@ -0,0 +1,169 @@
+
+    def _create_s3_presigned_url(self, document_reference: DocumentReference) -> str:
+        """Create a pre-signed URL for uploading a file"""
+        try:
+            response = self.s3_service.create_put_presigned_url(
+                document_reference.s3_bucket_name, document_reference.s3_file_key
+            )
+            logger.info(
+                f"Successfully created pre-signed URL for {document_reference.s3_file_key}"
+            )
+            return response
+        except ClientError as e:
+            logger.error(f"Failed to create pre-signed URL: {str(e)}")
+            raise FhirDocumentReferenceException(f"Failed to create pre-signed URL: {str(e)}")
+    
+    def _create_document_reference(
+        self,
+        nhs_number: str,
+        doc_type: SnomedCode,
+        fhir_doc: FhirDocumentReference,
+        current_gp_ods: str,
+        version: str,
+        s3_file_key: str,
+    ) -> DocumentReference:
+        """Create a document reference model"""
+        document_id = create_reference_id()
+
+        custodian = fhir_doc.custodian.identifier.value if fhir_doc.custodian else None
+        if not custodian:
+            custodian = (
+                current_gp_ods
+                if current_gp_ods not in PatientOdsInactiveStatus.list()
+                else PCSE_ODS_CODE
+            )
+        document_reference = DocumentReference(
+            id=document_id,
+            nhs_number=nhs_number,
+            current_gp_ods=current_gp_ods,
+            custodian=custodian,
+            s3_bucket_name=self.staging_bucket_name,
+            author=fhir_doc.author[0].identifier.value,
+            content_type=fhir_doc.content[0].attachment.contentType,
+            file_name=fhir_doc.content[0].attachment.title,
+            document_snomed_code_type=doc_type.code,
+            doc_status="preliminary",
+            status="current",
+            sub_folder="user_upload",
+            version=version,
+            s3_file_key=s3_file_key
+        )
+
+        return document_reference
+    
+    def _get_document_reference(self, document_id: str, table) -> DocumentReference:
+        documents = self.document_service.fetch_documents_from_table(
+            table=table,
+            search_condition=document_id,
+            search_key="ID",
+            query_filter=CurrentStatusFile,
+        )
+        if len(documents) > 0:
+            logger.info("Document found for given id")
+            return documents[0]
+        else:
+            raise FhirDocumentReferenceException(f"Did not find any documents for document ID {document_id}")
+
+    def _determine_document_type(self, fhir_doc: FhirDocumentReference) -> SnomedCode:
+        """Determine the document type based on SNOMED code in the FHIR document"""
+        if fhir_doc.type and fhir_doc.type.coding:
+            for coding in fhir_doc.type.coding:
+                if coding.system == SNOMED_URL:
```

Comment:
Will this fail on the first item if it's not a SNOMED_URL? What if the snomed code is correct but it's the 2nd item in the list?